adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1106,48 +1105,74 @@ protected final void handlePublishedMessage(final MqttPublishMessage message) {
-
-                            final ErrorSubscription errorSubscription = getErrorSubscription(context);
-                            final Future<Void> errorSentIfNeededFuture = Optional.ofNullable(errorSubscription)
-                                    .map(v -> publishError(errorSubscription, context, processing.cause(), span.context()))
-                                    .orElseGet(Future::succeededFuture);
-
-                            errorSentIfNeededFuture.onComplete(ar -> {
-                                final ErrorHandlingMode errorHandlingMode = context
-                                        .getErrorHandlingMode(errorSubscription != null);
-                                final boolean isTerminalError = false; // TODO implement detection of terminal error
-                                if (errorHandlingMode == ErrorHandlingMode.DISCONNECT || isTerminalError) {
-                                    if (context.deviceEndpoint().isConnected()) {
-                                        span.log("closing connection to device");
-                                        context.deviceEndpoint().close();
-                                    }
-                                } else if (context.isAtLeastOnce()) {
-                                    if (errorHandlingMode == ErrorHandlingMode.SKIP_ACK) {
-                                        span.log("skipped sending PUBACK");
-                                    } else if (context.deviceEndpoint().isConnected()) {
-                                        span.log("sending PUBACK");
-                                        context.acknowledge();
-                                    }
-                                }
-                                span.finish();
-                            });
                         }
+                        span.finish();
                     });
         }
 
+        private Future<Boolean> isTerminalError(final Throwable error, final String tenantId, final String deviceId,
+                final Device authenticatedDevice, final SpanContext spanContext) {
+
+            // if the device is not registered or disabled
+            if (error instanceof DeviceRegistrationException) {
+                // if the device is connected via a gateway
+                if (Objects.nonNull(authenticatedDevice) && !authenticatedDevice.getDeviceId().equals(deviceId)) {
+                    return getRegistrationAssertion(tenantId, authenticatedDevice.getDeviceId(), null, spanContext)
+                            .map(ok -> false)
+                            .recover(e -> {
+                                // if the gateway is not registered then it is a terminal error
+                                return Future.succeededFuture(e instanceof DeviceRegistrationException);
+                            });
+                }
+                // if the device is not connected via a gateway
+                return Future.succeededFuture(true);
```

Comment:
(The `if` in the comment is a bit confusing.)
I think the return statement should be changed to
````
return Future.succeededFuture(authenticatedDevice != null);
````
An *unauthenticated* client (which could potentially use a different deviceId for every telemetry/event message it publishes) should not be disconnected, FMPOV.

Reply:
IIRC anonymous devices are not supported. An unauthenticated device still should be registered, isn't it?  I see from the code that we don't allow unauthenticated devices that are not registered to send messages.

Reply:
FMPOV, the client that sent the MQTT CONNECT packet *without* authentication data doesn't necessarily represent just one device. The connected client will provide the tenant/device id in the topic path for each message that it publishes.
Therefore the client could be used for different devices, similar to a gateway publishing messages on behalf of different devices. That's why I wouldn't close the MQTT connection in that case.

Reply:
> FMPOV, the client that sent the MQTT CONNECT packet _without_ authentication data doesn't necessarily represent just one device.

Do you mean that it's an gateway? If so, then it should be registered, isn't it?


>  The connected client will provide the tenant/device id in the topic path for each message that it publishes.

In that case, the connected client (i.e gateway) should be registered, isn't it?

Reply:
I have updated and pushed the commit. Would you mind taking a look again?